File: packages/agents/coordination/src/priority.ts
/**
 * @create-something/agent-coordination
 *
 * Priority algorithms for multi-agent coordination.
 * Implements graph-based prioritization: Critical Path, Impact Score, PageRank-lite.
 */

import type { CoordinationDB, Issue, PriorityResult } from './types.js';
import { Tracker } from './tracker.js';

/**
 * Priority calculator for issue ordering.
 * Uses graph analysis to surface highest-impact work.
 */
export class Priority {
  private tracker: Tracker;

  constructor(private db: CoordinationDB) {
    this.tracker = new Tracker(db);
  }

  /**
   * Get prioritized list of ready issues.
   * Combines multiple signals: impact, priority level, age, dependencies.
   */
  async getPrioritized(limit: number = 10): Promise<PriorityResult[]> {
    // Get all ready issues (open, not blocked, not claimed)
    const ready = await this.tracker.getReadyIssues(100);

    if (ready.length === 0) {
      return [];
    }

    // Calculate scores for each issue
    const scored: PriorityResult[] = [];

    for (const issue of ready) {
      const score = await this.calculateScore(issue);
      scored.push(score);
    }

    // Sort by score descending
    scored.sort((a, b) => b.score - a.score);

    return scored.slice(0, limit);
  }

  /**
   * Calculate priority score for a single issue.
   * Higher score = should be worked on first.
   */
  async calculateScore(issue: Issue): Promise<PriorityResult> {
    const factors: { name: string; value: number; weight: number }[] = [];

    // 1. Priority level (P0=highest, P4=lowest)
    // Invert so P0 gets highest score
    const priorityScore = (4 - issue.priority) / 4;
    factors.push({ name: 'priority', value: priorityScore, weight: 0.3 });

    // 2. Impact: How many issues does this unblock?
    const impactScore = await this.calculateImpact(issue.id);
    factors.push({ name: 'impact', value: impactScore, weight: 0.35 });

    // 3. Age: Older issues get slight boost to prevent starvation
    const ageScore = this.calculateAgeScore(issue.createdAt);
    factors.push({ name: 'age', value: ageScore, weight: 0.1 });

    // 4. Connectivity: Is this a hub issue? (many connections)
    const connectivityScore = await this.calculateConnectivity(issue.id);
    factors.push({ name: 'connectivity', value: connectivityScore, weight: 0.15 });

    // 5. Project alignment: Issues in active projects get boost
    const projectScore = issue.projectId ? 0.5 : 0;
    factors.push({ name: 'project', value: projectScore, weight: 0.1 });

    // Calculate weighted score
    const totalScore = factors.reduce(
      (sum, f) => sum + f.value * f.weight,
      0
    );

    // Build reason string
    const topFactors = factors
      .filter((f) => f.value > 0.3)
      .sort((a, b) => b.value * b.weight - a.value * a.weight)
      .slice(0, 2)
      .map((f) => f.name);

    const reason = topFactors.length > 0
      ? `High ${topFactors.join(', ')}`
      : 'Default priority';

    return {
      issue,
      score: Math.round(totalScore * 100) / 100,
      reason,
    };
  }

  /**
   * Calculate impact score: How many issues does completing this unblock?
   * Uses recursive count of dependent issues.
   */
  async calculateImpact(issueId: string): Promise<number> {
    // Get all issues that this one blocks (directly and transitively)
    const blockedCount = await this.countBlockedIssues(issueId, new Set());

    // Normalize: 0 blocked = 0, 5+ blocked = 1
    return Math.min(blockedCount / 5, 1);
  }

  private async countBlockedIssues(issueId: string, visited: Set<string>): Promise<number> {
    if (visited.has(issueId)) return 0;
    visited.add(issueId);

    const { results } = await this.db
      .prepare(
        `SELECT d.to_id FROM dependencies d
         JOIN issues i ON i.id = d.to_id
         WHERE d.from_id = ? AND d.type = 'blocks'
         AND i.status IN ('open', 'blocked')`
      )
      .bind(issueId)
      .all<{ to_id: string }>();

    let count = results.length;

    // Recursively count transitively blocked issues
    for (const { to_id } of results) {
      count += await this.countBlockedIssues(to_id, visited);
    }

    return count;
  }

  /**
   * Calculate age score: Older issues get slight priority boost.
   * Prevents issue starvation.
   */
  private calculateAgeScore(createdAt: number): number {
    const now = Math.floor(Date.now() / 1000);
    const ageSeconds = now - createdAt;
    const ageDays = ageSeconds / (24 * 60 * 60);

    // Score increases with age, maxes out at 7 days
    return Math.min(ageDays / 7, 1);
  }

  /**
   * Calculate connectivity score: How connected is this issue in the graph?
   * Hub issues (many connections) are often important.
   */
  async calculateConnectivity(issueId: string): Promise<number> {
    const { results } = await this.db
      .prepare(
        `SELECT COUNT(*) as count FROM (
           SELECT from_id as id FROM dependencies WHERE to_id = ?
           UNION ALL
           SELECT to_id as id FROM dependencies WHERE from_id = ?
         )`
      )
      .bind(issueId, issueId)
      .all<{ count: number }>();

    const connectionCount = results[0]?.count ?? 0;

    // Normalize: 0 connections = 0, 10+ connections = 1
    return Math.min(connectionCount / 10, 1);
  }

  /**
   * Get the critical path: sequence of issues that, if delayed, delay everything.
   */
  async getCriticalPath(): Promise<Issue[]> {
    // Find issues with no dependents (end of chains)
    const { results: endpoints } = await this.db
      .prepare(
        `SELECT i.* FROM issues i
         WHERE i.status IN ('open', 'blocked', 'in_progress')
         AND NOT EXISTS (
           SELECT 1 FROM dependencies d WHERE d.from_id = i.id AND d.type = 'blocks'
         )`
      )
      .all<Record<string, unknown>>();

    if (endpoints.length === 0) {
      return [];
    }

    // Find the longest chain ending at each endpoint
    let longestPath: Issue[] = [];

    for (const endpoint of endpoints) {
      const path = await this.tracePath(endpoint.id as string, new Set());
      if (path.length > longestPath.length) {
        longestPath = path;
      }
    }

    return longestPath;
  }

  private async tracePath(issueId: string, visited: Set<string>): Promise<Issue[]> {
    if (visited.has(issueId)) return [];
    visited.add(issueId);

    const issue = await this.tracker.getIssue(issueId);
    if (!issue) return [];

    // Get all blockers of this issue
    const { results: blockers } = await this.db
      .prepare(
        `SELECT d.from_id FROM dependencies d
         JOIN issues i ON i.id = d.from_id
         WHERE d.to_id = ? AND d.type = 'blocks'
         AND i.status IN ('open', 'blocked', 'in_progress')`
      )
      .bind(issueId)
      .all<{ from_id: string }>();

    if (blockers.length === 0) {
      return [issue];
    }

    // Find longest path through blockers
    let longestPath: Issue[] = [];

    for (const { from_id } of blockers) {
      const path = await this.tracePath(from_id, new Set(visited));
      if (path.length > longestPath.length) {
        longestPath = path;
      }
    }

    return [...longestPath, issue];
  }

  /**
   * Get bottleneck issues: issues blocking the most other work.
   */
  async getBottlenecks(limit: number = 5): Promise<PriorityResult[]> {
    const { results } = await this.db
      .prepare(
        `SELECT i.*, COUNT(d.to_id) as blocked_count
         FROM issues i
         JOIN dependencies d ON d.from_id = i.id AND d.type = 'blocks'
         JOIN issues blocked ON blocked.id = d.to_id AND blocked.status IN ('open', 'blocked')
         WHERE i.status IN ('open', 'in_progress')
         GROUP BY i.id
         ORDER BY blocked_count DESC
         LIMIT ?`
      )
      .bind(limit)
      .all<Record<string, unknown> & { blocked_count: number }>();

    return results.map((row) => ({
      issue: this.rowToIssue(row),
      score: row.blocked_count,
      reason: `Blocking ${row.blocked_count} issue(s)`,
    }));
  }

  private rowToIssue(row: Record<string, unknown>): Issue {
    return {
      id: row.id as string,
      description: row.description as string,
      status: row.status as Issue['status'],
      projectId: row.project_id as string | null,
      parentId: row.parent_id as string | null,
      priority: row.priority as number,
      labels: JSON.parse((row.labels as string) || '[]'),
      metadata: JSON.parse((row.metadata as string) || '{}'),
      createdAt: row.created_at as number,
      updatedAt: row.updated_at as number,
      resolvedAt: row.resolved_at as number | null,
    };
  }
}

File: packages/loom/src/priority.rs
//! Priority Algorithms for Robot-Optimized Work Selection
//!
//! Implements graph-based prioritization: Impact Score, Critical Path, PageRank-lite.
//! Port of packages/agents/coordination/src/priority.ts for Rust/Loom.
//!
//! Usage:
//! ```rust,ignore
//! let priority = Priority::new(&work_store);
//! let ranked = priority.get_prioritized(10)?;
//! for result in ranked {
//!     println!("{}: {} (score: {:.2}, reason: {})", 
//!         result.task.id, result.task.title, result.score, result.reason);
//! }
//! ```

use crate::work::{Task, Status, WorkStore, WorkError, Dependency, DependencyType};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

/// Result of priority calculation for a single task
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PriorityResult {
    /// The task being scored
    pub task: Task,
    /// Computed priority score (0.0 - 1.0, higher = work on first)
    pub score: f64,
    /// Human-readable explanation of why this task is prioritized
    pub reason: String,
    /// Breakdown of scoring factors
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub factors: Vec<ScoringFactor>,
}

/// Individual factor contributing to priority score
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoringFactor {
    pub name: String,
    pub value: f64,
    pub weight: f64,
}

/// Critical path information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CriticalPath {
    /// Tasks in the critical path, ordered from start to end
    pub tasks: Vec<Task>,
    /// Total length of the path
    pub length: usize,
}

/// Priority calculator for work selection
pub struct Priority<'a> {
    store: &'a WorkStore,
}

impl<'a> Priority<'a> {
    /// Create a new priority calculator
    pub fn new(store: &'a WorkStore) -> Self {
        Self { store }
    }
    
    /// Get prioritized list of ready tasks
    /// Combines multiple signals: impact, priority level, age, dependencies
    pub fn get_prioritized(&self, limit: usize) -> Result<Vec<PriorityResult>, WorkError> {
        // Get all ready tasks
        let ready = self.store.ready()?;

        if ready.is_empty() {
            return Ok(vec![]);
        }

        // Pre-fetch all tasks and dependencies once to avoid O(n^2) DB queries
        let graph = DependencyGraph::build(self.store)?;

        // Calculate scores for each task
        let mut scored: Vec<PriorityResult> = Vec::new();

        for task in ready {
            let result = self.calculate_score_with_graph(&task, &graph)?;
            scored.push(result);
        }

        // Sort by score descending
        scored.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));

        Ok(scored.into_iter().take(limit).collect())
    }
    
    /// Calculate priority score for a single task (public API, fetches deps on demand)
    /// Higher score = should be worked on first
    pub fn calculate_score(&self, task: &Task) -> Result<PriorityResult, WorkError> {
        let graph = DependencyGraph::build(self.store)?;
        self.calculate_score_with_graph(task, &graph)
    }

    /// Calculate priority score using a pre-built dependency graph (O(1) lookups)
    fn calculate_score_with_graph(&self, task: &Task, graph: &DependencyGraph) -> Result<PriorityResult, WorkError> {
        let mut factors: Vec<ScoringFactor> = Vec::new();

        // 1. Priority level (Critical=highest, Low=lowest)
        // Weight: 0.30
        let priority_value = match task.priority {
            crate::work::Priority::Critical => 1.0,
            crate::work::Priority::High => 0.75,
            crate::work::Priority::Normal => 0.5,
            crate::work::Priority::Low => 0.25,
        };
        factors.push(ScoringFactor {
            name: "priority".to_string(),
            value: priority_value,
            weight: 0.30,
        });

        // 2. Impact: How many issues does this unblock?
        // Weight: 0.35
        let impact_value = graph.calculate_impact(&task.id);
        factors.push(ScoringFactor {
            name: "impact".to_string(),
            value: impact_value,
            weight: 0.35,
        });

        // 3. Age: Older tasks get slight boost to prevent starvation
        // Weight: 0.10
        let age_value = Self::calculate_age_score(task.created_at);
        factors.push(ScoringFactor {
            name: "age".to_string(),
            value: age_value,
            weight: 0.10,
        });

        // 4. Connectivity: Is this a hub task? (many connections)
        // Weight: 0.15
        let connectivity_value = graph.calculate_connectivity(&task.id);
        factors.push(ScoringFactor {
            name: "connectivity".to_string(),
            value: connectivity_value,
            weight: 0.15,
        });

        // 5. Issue type bonus: Bugs get slight priority over features
        // Weight: 0.10
        let type_value = match task.issue_type {
            crate::work::IssueType::Bug => 0.8,
            crate::work::IssueType::Feature => 0.5,
            crate::work::IssueType::Task => 0.5,
            crate::work::IssueType::Epic => 0.3,
            crate::work::IssueType::Chore => 0.4,
        };
        factors.push(ScoringFactor {
            name: "type".to_string(),
            value: type_value,
            weight: 0.10,
        });

        // Calculate weighted score
        let total_score: f64 = factors.iter()
            .map(|f| f.value * f.weight)
            .sum();

        // Build reason string from top factors
        let mut top_factors: Vec<_> = factors.iter()
            .filter(|f| f.value > 0.3)
            .map(|f| (f.name.clone(), f.value * f.weight))
            .collect();
        top_factors.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        let reason = if !top_factors.is_empty() {
            let names: Vec<_> = top_factors.iter().take(2).map(|(n, _)| n.as_str()).collect();
            format!("High {}", names.join(", "))
        } else {
            "Default priority".to_string()
        };

        Ok(PriorityResult {
            task: task.clone(),
            score: (total_score * 100.0).round() / 100.0,
            reason,
            factors,
        })
    }

    /// Calculate age score: Older tasks get slight priority boost
    /// Prevents task starvation
    fn calculate_age_score(created_at: DateTime<Utc>) -> f64 {
        let now = Utc::now();
        let age_seconds = (now - created_at).num_seconds() as f64;
        let age_days = age_seconds / (24.0 * 60.0 * 60.0);

        // Score increases with age, maxes out at 7 days
        (age_days / 7.0).min(1.0)
    }
    
    /// Get the critical path: sequence of tasks that, if delayed, delay everything
    pub fn get_critical_path(&self) -> Result<CriticalPath, WorkError> {
        let all_tasks = self.store.list_all()?;
        let active_tasks: Vec<_> = all_tasks.iter()
            .filter(|t| matches!(t.status, Status::Ready | Status::Blocked | Status::Claimed))
            .collect();
        
        if active_tasks.is_empty() {
            return Ok(CriticalPath { tasks: vec![], length: 0 });
        }
        
        // Find endpoint tasks (no outgoing blocking dependencies)
        let endpoints: Vec<_> = active_tasks.iter()
            .filter(|t| {
                // Check if this task blocks any other active task
                !active_tasks.iter().any(|other| {
                    if other.id == t.id {
                        return false;
                    }
                    self.store.get_all_dependencies(&other.id)
                        .map(|deps| deps.iter().any(|d| d.depends_on == t.id && d.dep_type == DependencyType::Blocks))
                        .unwrap_or(false)
                })
            })
            .collect();
        
        // Find longest path ending at each endpoint
        let mut longest_path: Vec<Task> = vec![];
        
        for endpoint in endpoints {
            let path = self.trace_path(&endpoint.id, &mut HashSet::new())?;
            if path.len() > longest_path.len() {
                longest_path = path;
            }
        }
        
        let length = longest_path.len();
        Ok(CriticalPath { tasks: longest_path, length })
    }
    
    fn trace_path(&self, task_id: &str, visited: &mut HashSet<String>) -> Result<Vec<Task>, WorkError> {
        if visited.contains(task_id) {
            return Ok(vec![]);
        }
        visited.insert(task_id.to_string());
        
        let task = match self.store.get(task_id)? {
            Some(t) => t,
            None => return Ok(vec![]),
        };
        
        // Get all blockers of this task
        let deps = self.store.get_all_dependencies(task_id)?;
        let blockers: Vec<_> = deps.iter()
            .filter(|d| d.dep_type == DependencyType::Blocks)
            .collect();
        
        if blockers.is_empty() {
            return Ok(vec![task]);
        }
        
        // Find longest path through blockers
        let mut longest_blocker_path: Vec<Task> = vec![];
        
        for blocker in blockers {
            if let Some(blocker_task) = self.store.get(&blocker.depends_on)? {
                if !matches!(blocker_task.status, Status::Done | Status::Cancelled) {
                    let path = self.trace_path(&blocker.depends_on, &mut visited.clone())?;
                    if path.len() > longest_blocker_path.len() {
                        longest_blocker_path = path;
                    }
                }
            }
        }
        
        longest_blocker_path.push(task);
        Ok(longest_blocker_path)
    }
    
    /// Get bottleneck tasks: tasks blocking the most other work
    pub fn get_bottlenecks(&self, limit: usize) -> Result<Vec<PriorityResult>, WorkError> {
        let graph = DependencyGraph::build(self.store)?;
        let active_tasks: Vec<_> = graph.all_tasks.iter()
            .filter(|t| matches!(t.status, Status::Ready | Status::Claimed))
            .collect();

        let mut bottlenecks: Vec<(Task, usize)> = Vec::new();

        for task in active_tasks {
            let blocked_count = graph.count_blocked_tasks(&task.id, &mut HashSet::new());
            if blocked_count > 0 {
                bottlenecks.push((task.clone(), blocked_count));
            }
        }

        // Sort by blocked count descending
        bottlenecks.sort_by(|a, b| b.1.cmp(&a.1));

        Ok(bottlenecks.into_iter()
            .take(limit)
            .map(|(task, blocked_count)| PriorityResult {
                task,
                score: blocked_count as f64,
                reason: format!("Blocking {} task(s)", blocked_count),
                factors: vec![],
            })
            .collect())
    }
}

/// Pre-computed dependency graph for O(1) lookups during scoring.
/// Built once from the database, then used for all scoring operations.
struct DependencyGraph {
    all_tasks: Vec<Task>,
    /// Dependencies keyed by task_id (task_id -> deps where this task is the dependent)
    deps_by_task: HashMap<String, Vec<Dependency>>,
    /// Reverse index: depends_on -> tasks that depend on it (via Blocks)
    blocked_by: HashMap<String, Vec<String>>,
}

impl DependencyGraph {
    /// Build the graph with two DB queries (all tasks + all dependencies)
    fn build(store: &WorkStore) -> Result<Self, WorkError> {
        let all_tasks = store.list_all()?;
        let mut deps_by_task: HashMap<String, Vec<Dependency>> = HashMap::new();
        let mut blocked_by: HashMap<String, Vec<String>> = HashMap::new();

        for task in &all_tasks {
            let deps = store.get_all_dependencies(&task.id)?;
            for dep in &deps {
                if dep.dep_type == DependencyType::Blocks {
                    blocked_by.entry(dep.depends_on.clone())
                        .or_default()
                        .push(task.id.clone());
                }
            }
            deps_by_task.insert(task.id.clone(), deps);
        }

        Ok(Self { all_tasks, deps_by_task, blocked_by })
    }

    /// Calculate impact score: How many tasks does completing this unblock?
    fn calculate_impact(&self, task_id: &str) -> f64 {
        let blocked_count = self.count_blocked_tasks(task_id, &mut HashSet::new());
        // Normalize: 0 blocked = 0, 5+ blocked = 1
        (blocked_count as f64 / 5.0).min(1.0)
    }

    fn count_blocked_tasks(&self, task_id: &str, visited: &mut HashSet<String>) -> usize {
        if visited.contains(task_id) {
            return 0;
        }
        visited.insert(task_id.to_string());

        let mut count = 0;
        if let Some(dependents) = self.blocked_by.get(task_id) {
            for dependent_id in dependents {
                // Only count active tasks
                if let Some(task) = self.all_tasks.iter().find(|t| t.id == *dependent_id) {
                    if task.status == Status::Ready || task.status == Status::Blocked {
                        count += 1;
                        count += self.count_blocked_tasks(dependent_id, visited);
                    }
                }
            }
        }
        count
    }

    /// Calculate connectivity score: How connected is this task in the graph?
    fn calculate_connectivity(&self, task_id: &str) -> f64 {
        // Outgoing deps (things this task depends on)
        let outgoing = self.deps_by_task.get(task_id).map(|d| d.len()).unwrap_or(0);
        // Incoming deps (things that depend on this task)
        let incoming = self.blocked_by.get(task_id).map(|d| d.len()).unwrap_or(0);

        // Normalize: 0 connections = 0, 10+ connections = 1
        ((outgoing + incoming) as f64 / 10.0).min(1.0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::work::{CreateTask, Priority as TaskPriority, IssueType};
    
    #[test]
    fn test_priority_calculation() {
        let mut store = WorkStore::in_memory().unwrap();
        
        // Create some tasks
        let high_priority = store.create(CreateTask {
            title: "Critical bug fix".to_string(),
            priority: TaskPriority::Critical,
            issue_type: IssueType::Bug,
            ..Default::default()
        }).unwrap();
        
        let low_priority = store.create(CreateTask {
            title: "Nice to have feature".to_string(),
            priority: TaskPriority::Low,
            issue_type: IssueType::Feature,
            ..Default::default()
        }).unwrap();
        
        let priority = Priority::new(&store);
        let ranked = priority.get_prioritized(10).unwrap();
        
        assert_eq!(ranked.len(), 2);
        // High priority task should be first
        assert_eq!(ranked[0].task.id, high_priority.id);
        assert!(ranked[0].score > ranked[1].score);
    }
    
    #[test]
    fn test_impact_scoring() {
        let mut store = WorkStore::in_memory().unwrap();
        
        // Create a blocker and blocked tasks
        let blocker = store.create(CreateTask {
            title: "Blocker task".to_string(),
            ..Default::default()
        }).unwrap();
        
        let blocked1 = store.create(CreateTask {
            title: "Blocked 1".to_string(),
            ..Default::default()
        }).unwrap();
        
        let blocked2 = store.create(CreateTask {
            title: "Blocked 2".to_string(),
            ..Default::default()
        }).unwrap();
        
        // Add dependencies
        store.add_dependency(&blocked1.id, &blocker.id).unwrap();
        store.add_dependency(&blocked2.id, &blocker.id).unwrap();
        
        let priority = Priority::new(&store);
        let ranked = priority.get_prioritized(10).unwrap();
        
        // The blocker should have higher priority due to impact
        let blocker_result = ranked.iter().find(|r| r.task.id == blocker.id).unwrap();
        let impact_factor = blocker_result.factors.iter().find(|f| f.name == "impact").unwrap();
        assert!(impact_factor.value > 0.0);
    }
    
    #[test]
    fn test_critical_path() {
        let mut store = WorkStore::in_memory().unwrap();
        
        // Create a chain: task1 → task2 → task3
        let task1 = store.create(CreateTask {
            title: "Task 1".to_string(),
            ..Default::default()
        }).unwrap();
        
        let task2 = store.create(CreateTask {
            title: "Task 2".to_string(),
            ..Default::default()
        }).unwrap();
        
        let task3 = store.create(CreateTask {
            title: "Task 3".to_string(),
            ..Default::default()
        }).unwrap();
        
        store.add_dependency(&task2.id, &task1.id).unwrap();
        store.add_dependency(&task3.id, &task2.id).unwrap();
        
        let priority = Priority::new(&store);
        let critical_path = priority.get_critical_path().unwrap();
        
        // Critical path should be length 3 (or less if task1 is not blocked)
        assert!(critical_path.length <= 3);
    }
}
